api/app/src/external/commonwell/document/document-query.ts

```diff
@@ -1,40 +1,54 @@
-import { CommonwellError, DocumentQueryResponse } from "@metriport/commonwell-sdk";
-import { MedicalDataSource } from "../..";
-import { createOrUpdate } from "../../../command/medical/document/create-or-update";
+import { CommonwellError, DocumentQueryResponse, Document } from "@metriport/commonwell-sdk";
+import { PassThrough } from "stream";
+import * as AWS from "aws-sdk";
 import { updateDocQueryStatus } from "../../../command/medical/document/document-query";
-import { DocumentReference } from "../../../domain/medical/document-reference";
-import { DocumentReferenceModel } from "../../../models/medical/document-reference";
 import { Patient } from "../../../models/medical/patient";
+import { Organization } from "../../../models/medical/organization";
+import { Facility } from "../../../models/medical/facility";
 import { capture } from "../../../shared/notifications";
 import { oid } from "../../../shared/oid";
 import { Util } from "../../../shared/util";
 import { makeCommonWellAPI, organizationQueryMeta } from "../api";
 import { getPatientData, PatientDataCommonwell } from "../patient-shared";
-import { DocumentWithLocation, getFileName, toDomain } from "./shared";
+import { downloadDocument } from "./document-download";
+import { upsertDocumentToFHIRServer } from "../../fhir/document/save-document-reference";
+import { toFHIR as toFHIRDocRef } from "../../fhir/document";
+import { DocumentWithFilename } from "./shared";
+import { Config } from "../../../shared/config";
+import { createS3FileName } from "../../../shared/external";
+import { getFileName } from "./shared";
 
-export async function getDocuments({
+const s3client = new AWS.S3();
+
+/**
+ * This is likely to be a long-running function
+ */
+export async function queryDocuments({
   patient,
   facilityId,
 }: {
   patient: Patient;
   facilityId: string;
-}): Promise<DocumentReference[]> {
+}): Promise<void> {
   try {
-    const cwDocuments = await internalGetDocuments({ patient, facilityId });
+    if (Config.isSandbox()) {
+      return;
+    }
 
-    const documents = cwDocuments.map(toDomain(patient));
+    const { organization, facility } = await getPatientData(patient, facilityId);
 
-    // Commonwell does not consistently return the same ID for a given document, so we don't
-    // update CW's entries, we replace them
-    await DocumentReferenceModel.destroy({
-      where: {
-        cxId: patient.cxId,
-        patientId: patient.id,
-        source: MedicalDataSource.COMMONWELL,
-      },
+    const cwDocuments = await internalGetDocuments({ patient, organization, facility });
+
+    const docsS3Refs = await dowloadAndUploadDocsToS3({
+      patient,
+      facilityId,
+      documents: cwDocuments,
     });
 
-    return await createOrUpdate(patient, documents);
+    for (const doc of docsS3Refs) {
+      const FHIRDocRef = toFHIRDocRef(doc, organization, patient);
+      await upsertDocumentToFHIRServer(FHIRDocRef);
```

Comment:
If we're not going to do the above suggestion now - can we not do a bulk upsert?

300 docs === 300 db calls, which in theory could be done in 1

Reply:
OTOH, if one fails we might fail all... not sure if the FHIR server would treat a batch upsert as a logical transaction or not. Just something to consider, I don't know the answer to that.